drivers/nec/display/display.cr

```diff
@@ -0,0 +1,397 @@
+module Nec; end
+module Nec::Display; end
+
+require "placeos-driver/interface/powerable"
+require "placeos-driver/interface/muteable"
+require "placeos-driver/interface/switchable"
+
+class Nec::Display::All < PlaceOS::Driver
+  include Interface::Powerable
+  include Interface::AudioMuteable
+
+  enum Inputs
+    Vga          = 1
+    Rgbhv        = 2
+    Dvi          = 3
+    Hdmi_set     = 4
+    Video1       = 5
+    Video2       = 6
+    Svideo       = 7
+    Tuner        = 9
+    Tv           = 10
+    Dvd1         = 12
+    Option       = 13
+    Dvd2         = 14
+    Display_port = 15
+    Hdmi         = 17
+    Hdmi2        = 18
+    Hdmi3        = 130
+    Usb          = 135
+
+    def to_s : String
+      Nec::Display::All.format_value(self.value)
+    end
+  end
+  include PlaceOS::Driver::Interface::InputSelection(Inputs)
+
+  # Discovery Information
+  tcp_port 7142
+  descriptive_name "NEC Display"
+  generic_name :Display
+
+  default_settings({
+    volume_min: 0,
+    volume_max: 100
+  })
+
+  @volume_min : Int32 = 0
+  @volume_max : Int32 = 100
+
+  @target_input : Input? = nil
+  @target_audio : Audio? = nil
+  @input_double_check : PlaceOS::Driver::Proxy::Scheduler? = nil
+
+  DELIMITER = 0x0D_u8
+
+  def on_load
+    # Communication settings
+    queue.delay = 120.milliseconds
+    # 0x0D (<CR> carriage return \r)
+    transport.tokenizer = Tokenizer.new(Bytes[DELIMITER])
+    on_update
+  end
+
+  def on_update
+    @volume_min = setting(Int32, :volume_min)
+    @volume_max = setting(Int32, :volume_max)
+  end
+
+  def connected
+    schedule.every(50.seconds, true) do
+      do_poll
+    end
+  end
+
+  def disconnected
+    schedule.clear
+  end
+
+  def power(state : Bool)
+    data = "C203D6"
+    current = self[:power]?
+
+    if current
+      data += "0004" # 0004 = Power Off
+      do_send(MsgType::Command, data, name: "power", delay: 10.seconds, timeout: 10.seconds)
+
+      self[:power] = false
+      logger.debug { "-- NEC LCD, requested to power off" }
```

Comment:
When adding debug message like this, it's a good idea to place them before the event they're referring to. This makes debugging significantly easier as if there's an issue, the action that led to it will appear in logs before hand, rather than possibly being dropped due to an exception.